src/helpers/serializer.ts

```diff
@@ -0,0 +1,187 @@
+      where[2] instanceof fuego.firestore.DocumentReference &&
+      !where[3]
+    ) {
+      return [
+        ...where.slice(0, 2),
+        where[2].path,
+        { type: 'ref' },
+      ] as WhereItem<Doc>
+    }
+
+    return where
+  }
+
+  // Serializer funciton for DocumentSnapshot
+  private static serializeSnapshot(
+    snapshot: DocumentSnapshot
+  ): SerializerOptions {
+    console.log('Serializing snapshot', snapshot.ref.path)
+
+    this.snapshotCache[snapshot.ref.path] = snapshot
+
+    return {
+      type: 'snapshot',
+      path: snapshot.ref.path,
+    }
+  }
+
+  private static serializeNumberOrSnapshot(
+    value?: number | DocumentSnapshot
+  ): SerializerOptions | number | undefined {
+    if (!value) return undefined
+    if (typeof value === 'number') return value
+
+    return this.serializeSnapshot(value)
+  }
+
+  // Serializer function for query
+  public static serializeQuery<Data extends object = {}>(
+    query: CollectionQueryType<Data>,
+    fuego: Fuego
+  ): string {
+    const { where, startAt, endAt, startAfter, endBefore, ...rest } = query
+
+    return JSON.stringify({
+      where: where ? this.serializeWhere(where, fuego) : undefined,
+      startAt: this.serializeNumberOrSnapshot(startAt),
+      endAt: this.serializeNumberOrSnapshot(endAt),
+      startAfter: this.serializeNumberOrSnapshot(startAfter),
+      endBefore: this.serializeNumberOrSnapshot(endBefore),
+      ...rest,
+    })
+  }
+
+  // Deserializer function for where condition
+  private static deserializeWhere<Doc extends object = {}>(
+    where: WhereType<Doc>,
+    fuego: Fuego
+  ): WhereType<Doc> {
+    if (this.multipleConditions(where)) {
+      return where.map(w => this.deserializeWhere(w, fuego)) as WhereArray<Doc>
+    }
+
+    if (where[3]?.type === 'date' && typeof where[2] === 'string') {
+      return [...where.slice(0, 2), new Date(where[2]), where[3]] as WhereItem<
+        Doc
+      >
+    }
+
+    if (where[3]?.type === 'ref' && typeof where[2] === 'string') {
+      return [
+        ...where.slice(0, 2),
+        fuego.db.doc(where[2]),
+        where[3],
+      ] as WhereItem<Doc>
+    }
+
+    return where
+  }
+
+  // Deserializer function for document snapshots
+  private static deserializeSnapshot(
+    snapshot: SerializerOptions
+  ): DocumentSnapshot | undefined {
+    if (snapshot.type !== 'snapshot') return
+
+    console.log(
+      'Deserializing snapshot',
+      snapshot,
+      this.snapshotCache[snapshot.path]
+    )
+
+    return this.snapshotCache[snapshot.path]
+  }
+
+  private static deserializeNumberOrSnapshot(
+    value?: number | SerializerOptions
+  ): DocumentSnapshot | number | undefined {
+    if (!value) return undefined
+    if (typeof value === 'number') return value
+
+    return this.deserializeSnapshot(value)
+  }
+
+  // Deserializer function for query
+  public static deserializeQuery<Data extends object = {}>(
+    queryString: string,
+    fuego: Fuego
```

Comment:
The reason for passing the fuego is to ensure the same version of firebase is used, because there are differences in the typings. I had trouble using it with my project due to this.

Reply:
That should still be solved at the initialization step, with something like this: https://github.com/nandorojo/swr-firestore/issues/59#issuecomment-719950071

Reply:
I do use my own Fuego instance, but I ran into problems with DocumentReference not being available in different versions of firebase, or at least available in different locations. If you have any suggested changes to fix this then that is welcome, the sooner we can merge this the better.